File: AlgoAndDS_Java/src/algoandds/sorting/QuickSort.java
package algoandds.sorting;

public class QuickSort implements Sortable{

	public void sortArray(int[] array) {

		quickSort(array, 0, array.length-1);
	}

	private void quickSort(int[] array, int low, int high) {

		if(low < high) {
			int q = partition(array, low, high);
			quickSort(array, low, q-1);
			quickSort(array, q+1, high);
		}
	}

	private int partition(int[] array, int low, int high) {
		
		int pivot = array[high];
		int i = low-1;
		for(int j = low; j <= high-1; j++) {
			if(pivot > array[j]) {
				i++;
				swap(array, i, j);
			}
		}
		swap(array, i+1, high);
		return i+1;
	}

	private void swap(int[] array, int i, int j) {

		int temp = array[i];
		array[i] = array[j];
		array[j] = temp;
	}

	 /*public static void main(String[] args) {
	    int[] array = {3,1,7,9,6,5,4,8,2};
	    new QuickSort().sortArray(array);
	    for (int n : array) {
	        System.out.println(n);
	    }
	}*/
}

File: AlgoAndDS_Python/algo/sorting/quick_sort.py
'''
Created on 30-Mar-2017

@author: sachin
'''
from random import randint

def swap(arr, i, p_index):
    temp = arr[i]
    arr[i] = arr[p_index]
    arr[p_index] = temp


def partition(arr, start, end):
    pivot = arr[end]
    p_index = start
    for i in range(start, end):
        if arr[i] <= pivot:
            swap(arr, i, p_index)
            p_index+=1
    swap(arr, p_index, end)
    return p_index

def quick_sort(arr, start, end):
    if start < end:
        p_index = partition(arr, start, end)
        quick_sort(arr, start, p_index-1)
        quick_sort(arr, p_index + 1, end)
        
arr = []
for i in range(20):
    arr.append(randint(10,100))
print(arr)
quick_sort(arr, 0, len(arr)-1)
print()
print(arr)